rust/opendp-ffi/src/core.rs

```diff
@@ -220,76 +238,77 @@ pub struct FfiTransformation {
 }
 
 impl FfiTransformation {
-    pub fn new_from_types<ID: 'static + Domain, OD: 'static + Domain, IM: 'static + Metric, OM: 'static + Metric>(value: Transformation<ID, OD, IM, OM>) -> *mut FfiTransformation {
-        let input_glue = FfiMetricGlue::<ID, IM>::new();
-        let input_glue = unsafe { transmute(input_glue) };
-        let output_glue = FfiMetricGlue::<OD, OM>::new();
-        let output_glue = unsafe { transmute(output_glue) };
-        Self::new(input_glue, output_glue, value)
+    pub fn new<ID: 'static + Domain, OD: 'static + Domain, IM: Metric, OM: Metric>(
+        input_glue: FfiMetricGlue<FfiDomain, FfiMetric>,
+        output_glue: FfiMetricGlue<FfiDomain, FfiMetric>,
+        value: Transformation<ID, OD, IM, OM>,
+    ) -> Self {
+        FfiTransformation { input_glue, output_glue, value: util::into_box(value) }
     }
+}
 
-    pub fn new<ID: 'static + Domain, OD: 'static + Domain, IM: Metric, OM: Metric>(input_glue: FfiMetricGlue<FfiDomain, FfiMetric>, output_glue: FfiMetricGlue<FfiDomain, FfiMetric>, value: Transformation<ID, OD, IM, OM>) -> *mut FfiTransformation {
-        let value = util::into_box(value);
-        let ffi_transformation = FfiTransformation { input_glue, output_glue, value };
-        util::into_raw(ffi_transformation)
+impl<DI: 'static + Domain, DO: 'static + Domain, MI: 'static + Metric, MO: 'static + Metric> From<Transformation<DI, DO, MI, MO>> for FfiTransformation {
+    fn from(value: Transformation<DI, DO, MI, MO>) -> Self {
+        FfiTransformation::new(
+            unsafe { transmute(FfiMetricGlue::<DI, MI>::new()) },
+            unsafe { transmute(FfiMetricGlue::<DO, MO>::new()) },
+            value)
     }
 }
 
 #[no_mangle]
 pub extern "C" fn opendp_core__error_free(this: *mut FfiError) {
-    util::into_owned(this);
+    util::into_owned(this).expect("Attempted to free a null pointer to an error.");
```

Comment:
I was thinking about this too. It seemed like a really lengthy handshake to make everything take two calls to deallocate. Am I correct in thinking that users will never have to deallocate manually? Maybe the overhead is unnecessary. Open to both routes.


util::into_owned returns an option so that a more specialized message can be attached at call sites.

Reply:
Yeah it'd be a major pain if it took two calls. That shouldn't be the case though, the `FfiResult` is returned on the stack, and if there's just a NULL inside it, there'd be nothing else to deallocate.

That said, I'm remembering now that usually `free(NULL)` is an allowed no-op, so I say we just leave it as is until it becomes an issue.